src/ucp/core/ucp_ep.c

```diff
@@ -831,13 +857,24 @@ ucs_status_ptr_t ucp_ep_close_nb(ucp_ep_h ep, unsigned mode)
 
     UCS_ASYNC_BLOCK(&worker->async);
 
+    ep->flags |= UCP_EP_FLAG_CLOSED;
     request = ucp_ep_flush_internal(ep,
                                     (mode == UCP_EP_CLOSE_MODE_FLUSH) ?
                                     UCT_FLUSH_FLAG_LOCAL : UCT_FLUSH_FLAG_CANCEL,
                                     NULL, 0, NULL,
                                     ucp_ep_close_flushed_callback, "close");
+
     if (!UCS_PTR_IS_PTR(request)) {
-        ucp_ep_disconnected(ep, mode == UCP_EP_CLOSE_MODE_FORCE);
+        if (ucp_ep_is_cm_local_connected(ep)) {
+            /* lanes already flushed, start disconnect on CM lane */
+            ucp_ep_cm_disconnect_cm_lane(ep);
+            close_req = ucp_ep_cm_close_request_get(ep);
+            ucp_ep_set_close_request(ep, close_req, "close");
```

Comment:
is it possible that the endpoint is not remote connected?
i would expect this to be the same code as ucp_ep_close_flushed_callback lines 824-831

Reply:
also, we should probably check for close_req!=NULL before using ucp_ep_set_close_request()

Reply:
> is it possible that the endpoint is not remote connected?

in this case EP is not locally connected as well.
this flush immediate completion means that flush which was initiated by disconnect callback is completed too, its completion callback initiates local disconnect and reset UCP_EP_FLAG_LOCAL_CONNECTED.

> also, we should probably check for close_req!=NULL before using ucp_ep_set_close_request()

I think better to have NULL-pointer request and valid state vs invalid state in protocol. I'll add NULL-check and error print where access it.

Lesser evil from 2, IMO. We are returning NO_MEMORY error here which usually means that normal workability of app is not possible from this point.

Reply:
added status_ptr instead of request in case of error + its print on access